src/mito2/src/worker/handle_truncate.rs

```diff
@@ -0,0 +1,51 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Handling truncate related requests.
+
+use common_query::Output;
+use common_telemetry::info;
+use store_api::logstore::LogStore;
+use store_api::storage::RegionId;
+
+use crate::error::{RegionNotFoundSnafu, Result};
+use crate::manifest::action::{RegionMetaAction, RegionMetaActionList, RegionTruncate};
+use crate::worker::RegionWorkerLoop;
+
+impl<S: LogStore> RegionWorkerLoop<S> {
+    pub(crate) async fn handle_truncate_request(&mut self, region_id: RegionId) -> Result<Output> {
```

Comment:
If there are running compaction tasks during truncating, it might commit SSTs to the truncated region. We might need to track compacting regions like the `FlushScheduler` or find out other ways to fix this. Consider adding a TODO for that.

Reply:
Can we add conditional logic in `handle_flush_finished` to check whether a region truncate occurred during the flush?
```rust
let version_data = region.version_control.current();
if version_data.version.flushed_entry_id >= request.flushed_entry_id {
    for file_meta in request.file_metas {
        region.file_purger.send_request(PurgeRequest {
            region_id: request.region_id,
            file_id: file_meta.file_id,
        })
    }
}
```

Reply:
Yes, we should check this. I think we could consider the necessity to add a truncated_entry_id for this check.

Reply:
Can't we use the updated `flushed_entry_id` from truncate region for checking?

https://github.com/GreptimeTeam/greptimedb/blob/1dd10f7aee28e71e81c407c70ee398e476746a01/src/mito2/src/region/opener.rs#L150-L153
Why not update the `flushed_entry_id` in the `version` to match the `flushed_entry_id` in the `RegionManifest`? This results in the `flushed_entry_id = 0` in the `version`  every time the region is opened.

https://github.com/GreptimeTeam/greptimedb/blob/1dd10f7aee28e71e81c407c70ee398e476746a01/src/mito2/src/region/opener.rs#L176-L178
Similarly, should the `last_entry_id` here also start from the `flushed_entry_id`?

Reply:
> Can't we use the updated flushed_entry_id from truncate region for checking?

Yes, we can use the flushed entry id currently. But this assumes that `version_data.version.flushed_entry_id >= request.flushed_entry_id` always implies the region is truncating. So I'm considering adding a `truncated_entry_id: Option<EntryId>`. We can persist it instead of `flushed_entry_id` and set `flushed_entry_id` to `truncated_entry_id` if it is `Some`.

> Why not update the flushed_entry_id in the version to match the flushed_entry_id in the RegionManifest? 

I already fixed it in this [PR](https://github.com/GreptimeTeam/greptimedb/pull/2343/files#diff-ef2e61d10144d024f82a4f23f7945356a9bc21963f968059fde1077ffc7c62f2)

There are still some PR works in progress.

> Similarly, should the last_entry_id here also start from the flushed_entry_id

It should start from flushed_entry_id for correctness. I'll also address it in another PR